utilities/transactions/lock/range/range_tree/range_tree_lock_manager.cc

```diff
@@ -0,0 +1,572 @@
+                                       const range_buffer& buffer, void*) {
+  auto txn = (PessimisticTransaction*)txnid;
+
+  RangeLockList* trx_locks =
+      ((RangeTreeLockTracker*)&txn->GetTrackedLocks())->getList();
+
+  MutexLock l(&trx_locks->mutex_);
+  if (trx_locks->releasing_locks_) {
+    /*
+      Do nothing. The transaction is releasing its locks, so it will not care
+      about having a correct list of ranges. (In TokuDB,
+      toku_db_txn_escalate_callback() makes use of this property, too)
+    */
+    return;
+  }
+
+  uint32_t cf_id = (uint32_t)lt->get_dict_id().dictid;
+
+  auto it = trx_locks->buffers_.find(cf_id);
+  it->second->destroy();
+  it->second->create();
+
+  toku::range_buffer::iterator iter(&buffer);
+  toku::range_buffer::iterator::record rec;
+  while (iter.current(&rec)) {
+    it->second->append(rec.get_left_key(), rec.get_right_key());
+    iter.next();
+  }
+}
+
+RangeTreeLockManager::~RangeTreeLockManager() {
+  autovector<void*> local_caches;
+  ltree_lookup_cache_->Scrape(&local_caches, nullptr);
+  for (auto cache : local_caches) {
+    delete static_cast<LockTreeMap*>(cache);
+  }
+
+  for (auto it : ltree_map_) {
+    ltm_.release_lt(it.second);
+  }
+  ltm_.destroy();
+}
+
+RangeLockManagerHandle::Counters RangeTreeLockManager::GetStatus() {
+  LTM_STATUS_S ltm_status_test;
+  ltm_.get_status(&ltm_status_test);
+  Counters res;
+
+  // Searching status variable by its string name is how Toku's unit tests
+  // do it (why didn't they make LTM_ESCALATION_COUNT constant visible?)
+  // lookup keyname in status
+  for (int i = 0; i < LTM_STATUS_S::LTM_STATUS_NUM_ROWS; i++) {
+    TOKU_ENGINE_STATUS_ROW status = &ltm_status_test.status[i];
+    if (strcmp(status->keyname, "LTM_ESCALATION_COUNT") == 0) {
+      res.escalation_count = status->value.num;
+      continue;
+    }
+    if (strcmp(status->keyname, "LTM_SIZE_CURRENT") == 0) {
+      res.current_lock_memory = status->value.num;
+    }
+  }
+  return res;
+}
+
+void RangeTreeLockManager::AddColumnFamily(const ColumnFamilyHandle* cfh) {
+  uint32_t column_family_id = cfh->GetID();
+
+  InstrumentedMutexLock l(&ltree_map_mutex_);
+  if (ltree_map_.find(column_family_id) == ltree_map_.end()) {
+    DICTIONARY_ID dict_id = {.dictid = column_family_id};
+    toku::comparator cmp;
+    cmp.create(compare_dbt_endpoints, (void*)cfh->GetComparator());
+    toku::locktree* ltree = ltm_.get_lt(dict_id, cmp,
+                                        /* on_create_extra*/ nullptr);
+    // This is ok to because get_lt has copied the comparator:
+    cmp.destroy();
+    ltree_map_.emplace(column_family_id, ltree);
+  } else {
+    // column_family already exists in lock map
+    // assert(false);
+  }
+}
+
+void RangeTreeLockManager::RemoveColumnFamily(const ColumnFamilyHandle* cfh) {
+  uint32_t column_family_id = cfh->GetID();
+  // Remove lock_map for this column family.  Since the lock map is stored
```

Comment:
it's a map from `uint32_t` (cf id) to `locktree*`, which shared ptr are you referring to?